include/ginkgo/core/base/lin_op.hpp

```diff
@@ -828,6 +828,31 @@ public:                                                                      \
                   "semi-colon warnings")
 
 
+/**
+ * Defines setters and getters for the parameters of a class, simplifying its
+ * construction by removing the repetitive typing of boiler plate code.
+ *
+ * @param _concrete_parameter_name  the parameter whose setter and getter is to
+ * be defined
+ *
+ * @ingroup LinOp
+ */
+#define GKO_ENABLE_SET_GET_PARAMETERS(_concrete_parameter_name, _type)       \
+public:                                                                      \
+    void set_##_concrete_parameter_name(_type other)                         \
+    {                                                                        \
+        this->_concrete_parameter_name##_ = other;                           \
+    }                                                                        \
+                                                                             \
+    _type get_##_concrete_parameter_name() const                             \
+    {                                                                        \
+        return this->_concrete_parameter_name##_;                            \
+    }                                                                        \
```

Comment:
Could we get by without specifying the type explicitly?
```suggestion
#define GKO_ENABLE_SET_GET_PARAMETERS(_concrete_parameter_name)              \
public:                                                                      \
    void set_##_concrete_parameter_name(                                     \
        decltype(this->_concrete_parameter_name##_) other)                   \
    {                                                                        \
        this->_concrete_parameter_name##_ = other;                           \
    }                                                                        \
                                                                             \
    auto get_##_concrete_parameter_name() const                              \
        -> decltype(this->_concrete_parameter_name##_)                       \
    {                                                                        \
        return this->_concrete_parameter_name##_;                            \
    }                                                                        \
```

Reply:
I had also tried this but it failed with 
```
/home/public_ginkgo/include/ginkgo/core/base/lin_op.hpp:858:18: error: invalid use of ‘this’ at top level
                                 decltype(this->_concrete_parameter_name##_) other)                   \ 
```

Additonally, there were [other places](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/stop/criterion.hpp#L109) where we pass in the type as well, so I assumed it was okay.

Reply:
Did you try leaving out the `this->` prefixes? That seems to work with the compilers I tested.

Reply:
You mean like `decltype(_concrete_parameter_name##_) other` ? Then it does not know that the `concrete_paramater_name##_` is a member variable right ? I do get errors saying that it is not declared.

Reply:
As far as I can tell, that's an issue of declaration order:
```cpp

class c {
    // works
    int i;
    auto f(decltype(i) a) -> decltype(i) {
        i = a;
    }
};

class c {
    // fails
    auto f(decltype(i) a) -> decltype(i) {
        i = a;
    }
    int i;
};
```

So if you put the getter/setter declaration after the member variable declaration, it should work

Reply:
Yeah. I guess you could declare the member variables earlier, but I am not sure I am too happy about that.

Reply:
(or declare the setters later, but I am also not happy about any of those approaches)